### ISSUE TITLE
Value of constant kernel must be scalar

### ISSUE THREAD

**Issue by femtomc** [dfm/tinygp#140]
Hi -- I'm using constant kernels inside of models whose logpdfs I eventually end up `vmap`ing (in importance sampling, for instance). In models where I parametrize `tinygp` kernels with random draws, one issue that I've encountered when attempting to `vmap`:

```
ValueError: The value of a constant kernel must be a scalar
```

It seems like it's "okay" if the values of constant kernels are JAX arrays.

I think we should define all kernels as `Pytree` implementors, and change the functionality so that tensors of any shape and size (under broadcast semantics) work under evaluation. I'm not sure what assumptions this change breaks.

In general, this will better facilitate usage of `tinygp` in probabilistic programming libraries. Right now, the above is sort of a breaking point for me which will prevent me from using `tinygp` with my library.

I'd love for this to change, and I would love to contribute to allow this change to happen -- as competitors (e.g. GPJax) also don't seem to implement kernels as `Pytree` implementors (I may be mistaken, but a cursory search through their codebase seemed not to turn up any `Pytree` node registration or metaclasses).

**Comment by femtomc** [comment]
The issue here: https://github.com/dfm/tinygp/blob/6684d95667f5139660f9dde4b2bb1a35994d9674/src/tinygp/kernels/base.py#L201-L222 is how checks like `jnp.ndim` will handle broadcasting under `vmap`. In normal use cases outside of PPLs, I suspect you'd never actually run into this question. 

But if I'm creating parameters under other random draws -- and then attempting to perform parallel logpdf evaluations -- you immediately run into this problem.

Have you thought about this issue at all?

**Comment by dfm** [comment]
The way the tinygp kernel design works is that the output of `evaluate` *must* be a scalar so that check is correct. But I also regularly vmap my models, so I'm confident that we can work out something that works in your use case without major issues. Can you share some specific example code demonstrating the kind of model you're trying to build?

**Comment by femtomc** [comment]
@dfm For whatever reason, when I fixed my other issue (#139) - it also fixed this one. I'm not exactly sure why yet, but I'm happy to close this for now.

**Comment by femtomc** [comment]
@dfm

**Comment by femtomc** [comment]
If my suspicion is correct, I'm not exactly sure how to fix this without changing these checks to occur at a different time.

What if, instead of at `Pytree` unflatten time -- the checks just occurred in `evaluate` (?) So e.g. in `evaluate`, if the user has correctly constructed the kernel -- we should see a `Traced<float32[]> with BatchTracer...` for parameters, for example.

Then, we can perform these checks in evaluate -- but don't perform them in `Pytree` post init because it might conflict with how trace-based systems save information about random variable invocations?

Not sure -- wdyt?

**Comment by dfm** [comment]
Interesting! I think that moving the check should be fine, but I'm not convinced that it would fix the problem. Do you think you can construct a simpler example that doesn't depend on your library that reproduces this issue so that we can iterate?

**Comment by femtomc** [comment]
Yes — I’ll create a simple example! It should help me figure out if it’s really the problem I think it is.

**Comment by femtomc** [comment]
@dfm Here is a notebook gist which showcases what I'm describing:

https://gist.github.com/femtomc/9d8299d514fed7da5290112949f4fa6a

The key issue is that when a GP or kernel instance is returned across the JAX API boundary -- unflattening back into a GP or kernel instance causes the check to run, and fail -- even if evaluation inside of the API boundary is acceptable (e.g. here, inside of `vmap` -- across the leading axis, evaluation uses a scalar).

**Comment by femtomc** [comment]
@dfm Btw -- if you find yourself busy, I can setup a PR myself with a little conversation about the checking invariants that you want to maintain for abstract kernel classes (like `Stationary`, etc).

**Comment by dfm** [comment]
Thanks! This is a very useful example. I'm at a conference this week so I won't be able to say much in the short term. I'm happy for you to start drafting a PR or we can just chat about it some more next week.

**Comment by femtomc** [comment]
I think this is closed by #148! Closing now, will re-open if I notice the same issue occurring elsewhere.
